tools/pythonpkg/src/pandas_analyzer.cpp

```diff
@@ -0,0 +1,387 @@
+		return true;
+	}
+	// Non-nested -> Nested is not possible
+	if (a < LogicalTypeId::STRUCT) {
+		return false;
+	}
+	// STRUCT -> LIST is not possible
+	if (b == LogicalTypeId::LIST) {
+		return false;
+	}
+	return true;
+}
+
+//@return Whether the two logicaltypes are compatible
+static bool CheckTypeCompatibility(const LogicalType &left, const LogicalType &right) {
+	return SameTypeRealm(left.id(), right.id());
+}
+
+struct StructToMapConvertData {
+	LogicalType map_value_type = LogicalType::SQLNULL;
+	bool is_valid_map = true;
+};
+
+//@return true if the struct is valid
+//@param valid_map Whether the format of the struct can be upgraded to map or not
+static bool IsStructColumnValid(const LogicalType &left, const LogicalType &right, StructToMapConvertData &data) {
+	//! Whether this is a valid struct
+	bool valid = true;
+
+	D_ASSERT(left.id() == LogicalTypeId::STRUCT && left.id() == right.id());
+
+	//! Child types of the two structs
+	auto &left_children = StructType::GetChildTypes(left);
+	auto &right_children = StructType::GetChildTypes(right);
+
+	if (left_children.size() != right_children.size()) {
+		valid = false;
+	}
+	//! Compare keys of struct case-insensitively
+	auto compare = CaseInsensitiveStringEquality();
+	for (idx_t i = 0; i < left_children.size(); i++) {
+		auto &left_child = left_children[i];
+		auto &right_child = right_children[i];
+
+		// keys in left and right don't match - upgrade to MAP
+		if (!compare(left_child.first, right_child.first)) {
+			valid = false;
+		}
+		// If values aren't compatible as structs, they wont be compatible as map either
+		if (!CheckTypeCompatibility(left_child.second, right_child.second)) {
+			data.is_valid_map = false;
+			return false;
+		}
+		if (!UpgradeType(data.map_value_type, left_child.second)) {
+			data.is_valid_map = false;
+			//! Could still be a valid struct, so we don't return here
+		}
+	}
+	return valid;
+}
+
+static LogicalType ConvertStructToMap(StructToMapConvertData &data) {
+	child_list_t<LogicalType> children;
+	// TODO: find a way to figure out actual type of the keys, not just the converted one
+	children.push_back(make_pair("key", LogicalType::LIST(LogicalType::VARCHAR)));
+	children.push_back(make_pair("value", LogicalType::LIST(data.map_value_type)));
+	return LogicalType::MAP(move(children));
+}
+
+static bool UpgradeType(LogicalType &left, const LogicalType &right) {
+	bool compatible = CheckTypeCompatibility(left, right);
+	if (!compatible) {
+		return false;
+	}
+	if (left.id() == LogicalTypeId::STRUCT && right.id() == left.id()) {
+		StructToMapConvertData convert_data;
+		if (!IsStructColumnValid(left, right, convert_data)) {
+			if (convert_data.is_valid_map) {
+				left = ConvertStructToMap(convert_data);
+			} else {
+				return false;
+			}
+		}
+	}
+	left = LogicalType::MaxLogicalType(left, right);
+	return true;
+}
+
+static py::object GetItem(py::handle &column, idx_t index) {
+	return column.attr("__getitem__")(index);
```

Comment:
Seems strange to directly call the dunder method explicitly here, does pybind not have something higher level that uses the index protocol under the hood?

Reply:
Hmm maybe, I hadn't considered this. But looking just now, it doesnt look like pybind11 has such a thing.

Reply:
I imagine the lookup cost for each call wouldn't be great either :/

Reply:
The `py::handle` seems to have the `[]` operator overloaded to call [`__getitem__`](https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L89), perhaps that can be used?

The lookup cost indeed will be slow but that is not a big problem as this only happens during the bind.